File: src/java/TravisAES.java
import lombok.extern.slf4j.Slf4j;

import javax.crypto.*;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import javax.xml.bind.DatatypeConverter;
import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.spec.KeySpec;

/**
 * @author travistran
 */
@Slf4j
public class TravisAES {

    private static final String AES = "AES";
    private static final String CIPHER_ALGORITHM = "AES/CBC/PKCS5Padding";
    private static final String KEY_ALGORITHM = "PBKDF2WithHmacSHA1";
    private static final int IV_SIZE = 128;
    private static final int IV_LENGTH = IV_SIZE / 4;
    private int keySize = 256;
    private int iterationCount = 1989;
    private DataTypeEnum dataType = DataTypeEnum.BASE64;
    private Cipher cipher;
    private int saltLength;

    public TravisAES() {
        try {
            cipher = Cipher.getInstance(CIPHER_ALGORITHM);
            saltLength = this.keySize / 4;
        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
            log.error(e.getMessage(), e);
        }
    }

    public TravisAES(int keySize, int iterationCount) {
        this.keySize = keySize;
        this.iterationCount = iterationCount;
        try {
            cipher = Cipher.getInstance(CIPHER_ALGORITHM);
            saltLength = this.keySize / 4;
        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
            log.error(e.getMessage(), e);
        }
    }

    private static byte[] generateRandom(int length) {
        SecureRandom random = new SecureRandom();
        byte[] randomBytes = new byte[length];
        random.nextBytes(randomBytes);
        return randomBytes;
    }

    private static byte[] fromBase64(String str) {
        return DatatypeConverter.parseBase64Binary(str);
    }

    private static String toBase64(byte[] ba) {
        return DatatypeConverter.printBase64Binary(ba);
    }

    private static byte[] fromHex(String str) {
        return DatatypeConverter.parseHexBinary(str);
    }

    private static String toHex(byte[] ba) {
        return DatatypeConverter.printHexBinary(ba);
    }

    public String encrypt(String salt, String iv, String passPhrase, String plainText) {
        try {
            SecretKey key = generateKey(salt, passPhrase);
            byte[] encrypted = doFinal(Cipher.ENCRYPT_MODE, key, iv, plainText.getBytes(StandardCharsets.UTF_8));
            String cipherText;
            if (dataType.equals(DataTypeEnum.HEX)) {
                cipherText = toHex(encrypted);
            } else {
                cipherText = toBase64(encrypted);
            }
            return cipherText;
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    public String encrypt(String passphrase, String plainText) {
        try {
            String salt = toHex(generateRandom(keySize / 8));
            String iv = toHex(generateRandom(IV_SIZE / 8));
            String cipherText = encrypt(salt, iv, passphrase, plainText);
            return salt + iv + cipherText;
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    public String decrypt(String salt, String iv, String passPhrase, String cipherText) {
        try {
            SecretKey key = generateKey(salt, passPhrase);
            byte[] encrypted;
            if (dataType.equals(DataTypeEnum.HEX)) {
                encrypted = fromHex(cipherText);
            } else {
                encrypted = fromBase64(cipherText);
            }
            byte[] decrypted = doFinal(Cipher.DECRYPT_MODE, key, iv, encrypted);
            return new String(decrypted, StandardCharsets.UTF_8);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    public String decrypt(String passPhrase, String cipherText) {
        try {
            String salt = cipherText.substring(0, saltLength);
            String iv = cipherText.substring(saltLength, saltLength + IV_LENGTH);
            String ct = cipherText.substring(saltLength + IV_LENGTH);
            return decrypt(salt, iv, passPhrase, ct);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    private byte[] doFinal(int encryptMode, SecretKey key, String iv, byte[] bytes) {
        try {
            cipher.init(encryptMode, key, new IvParameterSpec(fromHex(iv)));
            return cipher.doFinal(bytes);
        } catch (InvalidKeyException | InvalidAlgorithmParameterException | IllegalBlockSizeException |
                 BadPaddingException e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    private SecretKey generateKey(String salt, String passphrase) {
        try {
            SecretKeyFactory factory = SecretKeyFactory.getInstance(KEY_ALGORITHM);
            KeySpec spec = new PBEKeySpec(passphrase.toCharArray(), fromHex(salt), iterationCount, keySize);
            return new SecretKeySpec(factory.generateSecret(spec).getEncoded(), AES);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return null;
        }
    }

    public DataTypeEnum getDataType() {
        return dataType;
    }

    public void setDataType(DataTypeEnum dataType) {
        this.dataType = dataType;
    }

    public enum DataTypeEnum {
        HEX,
        BASE64,
    }
}

File: src/flutter/travis_aes.dart
import 'dart:convert';
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:pointycastle/block/aes_fast.dart';
import 'package:pointycastle/block/modes/cbc.dart';
import 'package:pointycastle/digests/sha1.dart';
import 'package:pointycastle/key_derivators/pbkdf2.dart';
import 'package:pointycastle/macs/hmac.dart';
import 'package:pointycastle/padded_block_cipher/padded_block_cipher_impl.dart';
import 'package:pointycastle/paddings/pkcs7.dart';
import 'package:pointycastle/pointycastle.dart';

class TravisAES {
  static const int _keySize = 256;
  static const int _saltLength = _keySize ~/ 8;
  static const int _ivSize = 128;
  static const int _ivLength = _ivSize ~/ 8;
  static const int _iterationCount = 1989;
  DataTypeEnum _dataType = DataTypeEnum.base64;

  TravisAES();

  Uint8List _generateRandom(int length) {
    final secureRandom = math.Random.secure();
    final bytes = Uint8List(length);
    for (int i = 0; i < length; i++) {
      bytes[i] = secureRandom.nextInt(256);
    }
    return bytes;
  }

  PaddedBlockCipher _generateCipher(
      bool encryptMode, Uint8List key, Uint8List iv) {
    final blockCipher = CBCBlockCipher(AESFastEngine());
    final paddedCipher = PaddedBlockCipherImpl(PKCS7Padding(), blockCipher);
    paddedCipher.init(
      encryptMode,
      PaddedBlockCipherParameters(
          ParametersWithIV(KeyParameter(key), iv), null),
    );
    return paddedCipher;
  }

  KeyDerivator _generateKeyDerivator() {
    return PBKDF2KeyDerivator(HMac(SHA1Digest(), 64));
  }

  Uint8List _generateKey(String passphrase, String salt) {
    final keyDerivator = _generateKeyDerivator();
    keyDerivator
        .init(Pbkdf2Parameters(_fromHex(salt), _iterationCount, _keySize ~/ 8));
    return keyDerivator.process(utf8.encoder.convert(passphrase));
  }

  String encryptWithSalt(
      String salt, String iv, String passphrase, String plaintext) {
    final key = _generateKey(passphrase, salt);
    final cipher = _generateCipher(true, key, _fromHex(iv));
    final encrypted = cipher.process(utf8.encoder.convert(plaintext));
    if (_dataType == DataTypeEnum.hex) {
      return _toHex(encrypted);
    } else {
      return base64Encode(encrypted);
    }
  }

  String encrypt(String passphrase, String plaintext) {
    final salt = _toHex(_generateRandom(_saltLength));
    final iv = _toHex(_generateRandom(_ivLength));
    final cipherText = encryptWithSalt(salt, iv, passphrase, plaintext);
    return salt + iv + cipherText;
  }

  String decryptWithSalt(
      String salt, String iv, String passphrase, String cipherText) {
    final key = _generateKey(passphrase, salt);
    final cipher = _generateCipher(false, key, _fromHex(iv));
    final Uint8List encrypted;
    if (_dataType == DataTypeEnum.hex) {
      encrypted = _fromHex(cipherText);
    } else {
      encrypted = base64Decode(cipherText);
    }
    final decrypted = cipher.process(encrypted);
    return utf8.decode(decrypted);
  }

  String decrypt(String passphrase, String cipherText) {
    String salt = cipherText.substring(0, _saltLength * 2);
    String iv =
        cipherText.substring(_saltLength * 2, _saltLength * 2 + _ivLength * 2);
    String ct = cipherText.substring(_saltLength * 2 + _ivLength * 2);
    return decryptWithSalt(salt, iv, passphrase, ct);
  }

  String _toHex(Uint8List bytes) {
    return hex.encode(bytes);
  }

  Uint8List _fromHex(String hexString) {
    return Uint8List.fromList(hex.decode(hexString));
  }

  DataTypeEnum get dataType => _dataType;

  set dataType(DataTypeEnum dataType) => _dataType = dataType;
}

enum DataTypeEnum {
  hex,
  base64,
}
